src/syntax/syntax.rkt

```diff
@@ -30,81 +71,93 @@
   (with-handlers ([exn:fail?
                    (λ (e) (error 'constant-info "Unknown constant or field: ~a ~a"
                                                 constant field))])
-    (table-ref constants constant field)))
-
-(define (register-constant! constant name ctype attrib-dict)
-  (table-set! constants name
-              (make-hash (cons (cons 'type ctype) attrib-dict)))
+    (table-ref constant-impls constant field)))
+
+(define (dict-merge dict dict2)
+  (for/fold ([dict dict]) ([(key value) (in-dict dict2)])
+    (dict-set dict key value)))
+
+(define (register-constant-impl! constant name ctype attrib-dict)
+  (define default-attrib (table-ref-all constants constant))
+  (unless default-attrib
+    (error 'register-constant-impl! "Real constant does not exist: ~a" constant))
+  (define attrib-dict* (dict-merge default-attrib attrib-dict))
+  (table-set! constant-impls name
+              (make-hash (cons (cons 'type ctype) attrib-dict*)))
```

Comment:
This is looking more and more like just a custom implementation of struct subclassing...